File: sparktest/splitPointCloud/src/main/java/com/SplitUtils.java
package com;



public class SplitUtils {

    public static double getClod(long pointNum,long pointNumPerNode,int dimension){

        int maxLevel = getMaxLevel(pointNum,pointNumPerNode,dimension);
        double random = Math.random();
        double clod =  Math.log((Math.pow(2,dimension*maxLevel+dimension)-1)*random+1)/(dimension*Math.log(2));

        return clod;
    }

    public static double getClod(int maxLevel,int dimension){

        double random = Math.random();
        double clod = (Math.log((Math.pow(2,dimension*maxLevel+dimension)-1)*random+1)/(dimension*Math.log(2)));
        return clod;
    }



    public static int getMaxLevel(long pointNum,long pointNumPerNode,int dimension){
        long blockNum = pointNum/pointNumPerNode;
        int maxLevel = (int)Math.ceil(Math.log(blockNum)/Math.log(Math.pow(2,dimension)));
        return maxLevel;
    }

    /**
     *
     * @param x
     * @param y
     * @param z
     * @param boundingBox maxx, maxy, maxz,minx,miny,minz
     * @param clod
     * @return
     */
    public static String getOctreeNodeName(double x,double y,double z,double[] boundingBox,double clod){
        int dlod = (int)clod;
        String xLocation = getLocationOnSingleAxis(x,boundingBox[0],boundingBox[3],dlod);
        String yLocation = getLocationOnSingleAxis(y,boundingBox[1],boundingBox[4],dlod);
        String zLocation = getLocationOnSingleAxis(z,boundingBox[2],boundingBox[5],dlod);

        return getNodeName(xLocation,yLocation,zLocation);

    }


    public static String getLocationOnSingleAxis(double x,double maxx,double minx,int dlod){
        String locationOnSingleAxis = "";
        /*
        if(dlod==0)
            return nodeName;

        double middlex = minx+(maxx-minx)/2;
        if(x<middlex)
            return "0"+getLocationOnSingleAxis(x,middlex,minx,dlod-1);
        else
            return "1"+getLocationOnSingleAxis(x,maxx,middlex,dlod-1);

         */

        if(maxx<minx)
            throw new IllegalArgumentException("输入范围最大最小值颠倒！");

        while (dlod>0){
            double middlex = minx+(maxx-minx)/2;
            if(x<middlex){
                locationOnSingleAxis+="0";
                maxx = middlex;
            }
            else{
                locationOnSingleAxis+="1";
                minx = middlex;
            }
           dlod--;
        }
        return locationOnSingleAxis;
    }

    public static String getNodeName(String xLocation,String yLocation,String zLocation){
        if(xLocation.length()==yLocation.length()?(xLocation.length()==zLocation.length()?false:true):true)
            throw new IllegalArgumentException("三个字符串参数长度不一致！");
        String nodeName = "r";
        int length = xLocation.length();
        for(int i=0;i<length;i++){
            String locationStr = ""+xLocation.charAt(i)+yLocation.charAt(i)+zLocation.charAt(i);
            nodeName += "" +Integer.parseInt(locationStr,2);
        }
        return nodeName;
    }


    public static double[] getXYZOffset(String nodeKey,double[] boundingBox){
        int level = nodeKey.length()-1;
        double xOffset = 0.0;
        double yOffset = 0.0;
        double zOffset = 0.0;
        double boxLength = boundingBox[0]-boundingBox[3];
        for(int i=1;i<level+1;i++){
            int childNodeIndex = Integer.valueOf(nodeKey.substring(i,i+1));
            xOffset += Integer.valueOf(childNodeIndex>>2&1)*boxLength/Math.pow(2,i);
            yOffset += Integer.valueOf(childNodeIndex>>1&1)*boxLength/Math.pow(2,i);
            zOffset += Integer.valueOf(childNodeIndex&1)*boxLength/Math.pow(2,i);
        }
        return new double[]{xOffset,yOffset,zOffset};
    }




}

File: sparktest/readLAS/src/splitUtils.py
# -*- encoding: utf-8 -*-
import numpy as np
import math


def getClod(maxLevel, dimension):
    random = np.random.rand()
    clod = (math.log((math.pow(2,dimension*maxLevel+dimension)-1)*random+1)/(dimension*math.log(2)))
    return clod


def getMaxLevel(pointNum, pointNumPerNode, dimension):
    blockNum = pointNum/pointNumPerNode
    maxLevel = int(math.ceil(math.log(blockNum)/math.log(math.pow(2,dimension))))
    return maxLevel


def getOctreeNodeKeyAndOffset(point, boundingBoxMax, boundingBoxMin, dlod):
    pointZIndex = ""
    pointMax = [boundingBoxMax[0], boundingBoxMax[1], boundingBoxMax[2]]
    pointMin = [boundingBoxMin[0], boundingBoxMin[1], boundingBoxMin[2]]
    offset = [boundingBoxMin[0], boundingBoxMin[1], boundingBoxMin[2]]
    while dlod>0:
        for i in range(3):
            mid = (pointMax[i]-pointMin[i])/2+pointMin[i]
            if(point[i]<mid):
                pointZIndex += "0"
                pointMax[i] = mid
            else:
                pointZIndex += "1"
                offset[i] += mid-pointMin[i]
                pointMin[i] = mid
        dlod = dlod-1
    nodekey = "r" + "".join([str(int(pointZIndex[i:i+3],2)) for i in range(len(pointZIndex)) if i%3==0])
    return nodekey,offset


'''低位在前'''
def pointInZigZagFormat(point):
    zigZagBytes = bytes()

    for i in range(3):
        x = point[i]
        while((x >> 7) > 0):

            zigZagBytes += bytes([(1 << 7)+ (x & 0x7f)])

            x = x >> 7
        zigZagBytes += bytes([x])
    return zigZagBytes


if __name__ == '__main__':
    point = [58098, 16397, 31411]
    print(pointInZigZagFormat(point))
